evennia/contrib/cooldowns.py

```diff
@@ -0,0 +1,188 @@
+    supports DB attributes (like a Character or Account).
+
+    A cooldown is a timer that is usually used to limit how often
+    some action can be performed or some effect can trigger. When a
+    cooldown is first set, it counts down from the amount of time
+    provided back to zero, at which point it is considered ready again.
+
+    Cooldowns are named with an arbitrary string, and that string is used
+    to check on the progression of the cooldown. Each cooldown is tracked
+    separately and independently.
+
+    Cooldowns are saved persistently, so they survive reboots. This
+    module does not register or provide callback functionality for when
+    a cooldown becomes ready again. Users of cooldowns are expected to
+    query the state of any cooldowns they are interested in.
+
+    Methods:
+    - ready(name): Checks whether a given cooldown name is ready.
+    - time_left(name): Returns how much time is left on a cooldown.
+    - set(name, seconds): Sets a given cooldown to last for a certain
+        amount of time. Until then, ready() will return False for that
+        cooldown name.
+    - extend(name, seconds): Like set, but adds time to the given
+        cooldown name. If it doesn't exist yet, calling this is equivalent
+        to calling set.
+    - reset(cooldown): Resets a given cooldown, causing ready() to return
+        True for that cooldown immediately.
+    - clear(): Resets all cooldowns.
+    """
+
+    def __init__(self, obj, db_attribute="cooldowns"):
+        if not obj.attributes.has(db_attribute):
+            obj.attributes.add(db_attribute, {})
+
+        self.data = obj.attributes.get(db_attribute)
+        self.cleanup()
+
+    @property
+    def all(self):
+        """
+        Returns a list of all keys in this object.
+        """
+        return list(self.data.keys())
+
+    def ready(self, *args):
+        """
+        Checks whether all of the provided cooldowns are ready (expired).
+        If a requested cooldown does not exist, it is considered ready.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (bool): True if each cooldown has expired or does not exist.
+        """
+        return self.time_left(*args) <= 0
+
+    def time_left(self, *args):
+        """
+        Returns the maximum amount of time left on one or more given
+        cooldowns. If a requested cooldown does not exist, it is
+        considered to have 0 time left.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (int): Number of seconds until all provided cooldowns are
+                ready. Returns 0 if all cooldowns are ready (or don't
+                exist.)
+        """
+        now = time.time()
+        cooldowns = [self.data[x] - now for x in args if x in self.data]
+        if not cooldowns:
+            return 0
+        return math.ceil(max(max(cooldowns), 0))
```

Comment:
While sub-second precision does not make much sense for repeating tasks (echoing messages faster than 1/s is not really that useful to a user), it seems unnecessary to limit this check to second-precision. One of the main uses of cooldowns are for combat systems, and sub-second precision can make use of those (even though you don't need to echo it to the user that fast). 

`ceil` should allow the `.ready` to still operate at sub-second precision (it's 1 until it suddenly flips over to 0), but it's useful to be able to see exactly how much time is left (e.g. by other code using this). So I suggest changing this to return floats by default instead. 

You could add a kwarg `precision` or `use_int` or something maybe?

Reply:
Good point. I added a `use_int` kwarg (default False) and changed everything to be float-first.